### ISSUE TITLE
[feature request] Component-wise transform of elements for multi-variate distributions

### ISSUE THREAD

**Issue by ahmadsalim** [pytorch/pytorch#11755]
Hi Torch devs!

I am thinking of adding a PR with a component-wise transform for multi-variate distributions (under https://github.com/pytorch/pytorch/blob/master/torch/distributions/transforms.py).
I already have the code ready, since I am using it to transform Multivariate Normals in a pyro-ppl model.

I read in the contribution guide that potential PRs should be discussed here before. Would you welcome such as a contribution?

Thanks!
Ahmad Salim

cc @vincentqb @fritzo @neerajprad @alicanb @vishwakftw

**Comment by vishwakftw** [comment]
cc: @fritzo @alicanb @neerajprad

**Comment by alicanb** [comment]
Can you give an example?

**Comment by ahmadsalim** [comment]
Sure!
I have a model where some of the observed variables are angles and distances, which I want to correlate. So I am using a multivariate normal to represent a vector of an angle and a distance, and I transform the angle by wrapping and the distance by exponentiation (so I get a log-normal prior on the distance component).

More formally, let us assume that we have some vector of random variables `(x_1, ..., x_n)` which we want to apply invertable transforms `(t_1, ..., t_n)` component-wise to get a vector of transformed RVs `(y_1, ..., y_n)` (so, `y_i = t_i(x_i)` and `x_i = t_i^[-1](y_i)`. The transform I have implemented allows this to be done easily.

**Comment by fritzo** [comment]
To me the most natural way to compose transforms would be through classes `CatTransform` and `StackTransform` that mimic the shape behavior of `torch.cat` and `torch.stack`, including a `dim` kwarg. In particular these would play well with broadcasting for e.g. batching. @ahmadsalim I think `CatTransform` is pretty close to your component-wise transform.

**Comment by ahmadsalim** [comment]
@fritzo Thanks for the suggestion.
As far as I understand these are hypothetical transformations, right? 😄 

In that case, I can take a look whether to implement them.

**Comment by fritzo** [comment]
@ahmadsalim yes these `StackTransform` and `CatTransform` are hypothetical, but they are natural objects (functors) with clearly defined semantics:
```py
class CatTransform(Transform):
    def __init__(self, seq, dim=0):
        self.transforms = list(seq)
        assert all(isinstance(t, Transform) for t in self.transforms)
        ...

class StackTransform(Transform):
    def __init__(self, seq, dim=0):
        self.transforms = list(seq)
        assert all(isinstance(t, Transform) for t in self.transforms)
        ...

class TestFunctors(TestCase):
  def test_cat_transform(self):
    x1, x2, x3 = ...
    t1, t2, t3 = ...
    dim = ...
    x = torch.cat([x1, x2, x3], dim=dim)
    t = CatTransform([t1, t2, t3], dim=dim)
    actual = t(x)
    expected = torch.cat([t1(x1), t2(x2), t3(x3)], dim=dim)
    self.assertEqual(expected, actual)

  def test_stack_transform(self):
    x1, x2, x3 = ...
    t1, t2, t3 = ...
    dim = ...
    x = torch.stack([x1, x2, x3], dim=dim)
    t = StackTransform([t1, t2, t3], dim=dim)
    actual = t(x)
    expected = torch.stack([t1(x1), t2(x2), t3(x3)], dim=dim)
    self.assertEqual(expected, actual)
```

**Comment by colesbury** [comment]
Is this complete?

**Comment by fritzo** [comment]
This was completed in https://github.com/pytorch/pytorch/pull/11868
